File: workspace/runners/kotlin/engine/src/main/kotlin/io/bosca/api/Security.kt
package io.bosca.api

import io.bosca.graphql.*
import io.bosca.graphql.fragment.Group
import io.bosca.graphql.fragment.LoginResponse
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class Security(network: NetworkClient) : Api(network) {

    private val mutex = Mutex()
    var keepTokenUpdated = true
    var token: LoginResponse? = null

    suspend fun refreshToken() {
        mutex.withLock {
            val response = network.graphql.mutation(RefreshTokenMutation(token?.refreshToken ?: return)).execute()
            response.validate()
            token = response.data?.security?.login?.refreshToken?.loginResponse
            network.graphqlToken = token?.token?.token
        }
    }

    private suspend fun loginInternal(identifier: String, password: String) {
        val response = network.graphql.mutation(LoginMutation(identifier, password)).execute()
        response.validate()
        token = response.data?.security?.login?.password?.loginResponse
        network.graphqlToken = token?.token?.token
    }

    @OptIn(DelicateCoroutinesApi::class)
    private fun keepTokenUpdated(identifier: String, password: String) {
        GlobalScope.launch {
            while (keepTokenUpdated) {
                val expiresAt = (token?.token?.expiresAt ?: 240).toLong() * 1000
                val issuedAt = (token?.token?.issuedAt ?: 0).toLong() * 1000
                val delay = (expiresAt - issuedAt) - 120_000
                delay(delay)
                if (!keepTokenUpdated) return@launch
                try {
                    refreshToken()
                } catch (e: Exception) {
                    println("failed to update token: $e")
                    if (!keepTokenUpdated) return@launch
                    try {
                        loginInternal(identifier, password)
                    } catch (e: Exception) {
                        println("failed to login, waiting 30 seconds... :: $e")
                        delay(30_000)
                    }
                }
            }
        }
    }

    suspend fun login(identifier: String, password: String) {
        loginInternal(identifier, password)
        keepTokenUpdated(identifier, password)
    }

    suspend fun addPrincipalGroup(principalId: String, groupId: String) {
        val response = network.graphql.mutation(AddPrincipalGroupMutation(groupId, principalId)).execute()
        response.validate()
    }

    suspend fun addGroup(name: String, description: String) {
        val response = network.graphql.mutation(AddGroupMutation(name, description)).execute()
        response.validate()
    }

    suspend fun getGroups(offset: Int, limit: Int): List<Group> {
        val response = network.graphql.query(GetGroupsQuery(offset, limit)).execute()
        response.validate()
        return response.data?.security?.groups?.all?.map { it.group } ?: emptyList()
    }

    suspend fun getPermissionActions(): List<String> {
        val response = network.graphql.query(GetPermissionActionsQuery()).execute()
        response.validate()
        return response.data?.security?.actions?.map { it.uppercase() } ?: emptyList()
    }
}

File: workspace/web/administration/lib/bosca/security.ts
import { Api } from './api'
import {
  GetGroupsDocument,
  GetPrincipalsDocument,
  type GroupFragment,
  LoginDocument,
  type LoginResponse,
  type PrincipalFragment,
  type ProfileInput,
  SignUpDocument,
  VerifySignUpDocument,
} from '~/lib/graphql/graphql'
import { NetworkClient } from '~/lib/bosca/networkclient'
import type { AsyncData } from '#app/composables/asyncData'

export class Security<T extends NetworkClient> extends Api<T> {
  constructor(network: T) {
    super(network)
  }

  getGroups(
    offset: number | Ref<number, null>,
    limit: number | Ref<number, null>,
  ): AsyncData<Array<GroupFragment> | null, any> {
    return this.executeAndTransformAsyncData(
      GetGroupsDocument,
      { offset, limit },
      (data) => {
        if (!data) return null
        return data.security?.groups?.all as Array<GroupFragment>
      },
    )
  }

  getPrincipals(
    offset: number | Ref<number, null>,
    limit: number | Ref<number, null>,
  ): AsyncData<Array<PrincipalFragment> | null, any> {
    return this.executeAndTransformAsyncData(
      GetPrincipalsDocument,
      { offset, limit },
      (data) => {
        if (!data) return null
        return data.security?.principals?.all as Array<PrincipalFragment>
      },
    )
  }

  async loginWithPassword(
    identifier: string,
    password: string,
  ): Promise<LoginResponse> {
    const response = await this.network.execute(LoginDocument, {
      identifier,
      password,
    })
    if (!response) throw new Error('invalid response')
    // @ts-ignore: not sure why this is confused
    const login = response.security.login.password
    this.network.token = login.token.token
    return login
  }

  async verify(token: string): Promise<void> {
    await this.network.execute(VerifySignUpDocument, {
      token,
    })
  }

  async signUpWithPassword(
    profile: ProfileInput,
    identifier: string,
    password: string,
  ) {
    await this.network.execute(SignUpDocument, {
      profile,
      identifier,
      password,
    })
  }
}

// suspend fun login(identifier: String, password: String) {
//     val response = network.client.query(LoginQuery(identifier, password)).execute()
//     response.validate()
//     network.token = response.data?.security?.login?.password?.token?.token
// }
//
// suspend fun getGroups(): List<Group> {
//     val response = network.client.query(GetGroupsQuery()).execute()
//     response.validate()
//     return response.data?.security?.groups?.all?.map { it.group } ?: emptyList()
// }
//
// suspend fun getPermissionActions(): List<String> {
//     val response = network.client.query(GetPermissionActionsQuery()).execute()
//     response.validate()
//     return response.data?.security?.actions?.map { it.uppercase() } ?: emptyList()
// }
// }
